keps/sig-network/2030-endpointslice-subsetting/README.md

```diff
@@ -0,0 +1,267 @@
+```
+matchLabels:
+  endpointslice.kubernetes.io/for-zone: example-zone
+```
+
+#### Select EndpointSlices With Matching Region
+```
+matchLabels:
+  endpointslice.kubernetes.io/for-region: example-region
+```
+
+#### Kube-Proxy
+When the `EndpointSliceSubsetting` feature gate is set to true on `Kube-Proxy`,
+it will use these selectors to filter EndpointSlices.
+
+### Controller Implementation
+Although a controller implementation is out of scope for this KEP, it is worth
+discussing what that might look like. For reference, [KEP
+#2004](https://github.com/kubernetes/enhancements/issues/2004) discusses how this
+could be implemented for the EndpointSlice controller. That proposal involves 3
+potential approaches - Original, PreferZone, and RequireZone.
+
+None of the proposed approaches would involve data duplication. Each
+Pod/endpoint would continue to live in a single EndpointSlice. The reason they
+might end up with more EndpointSlices would be less efficient packing. Here's
+the number of EndpointSlices that would result based on the number of endpoints
+a Service has in a 3 zone cluster.
+
+| # endpoints | Original # slices | PreferZone # slices | RequireZone # slices |
+|-|-|-|-|
+| 6 | 1 | 1 | 3 |
+| 80 | 1 | 3 | 3 |
+| 250 | 3 | 3 | 3 |
+
+The RequireZone approach requires at least one EndpointSlice per zone per
+Service. The PreferZone also has the same requirement unless the minimum
+threshold has been met. Before that threshold, a single shared EndpointSlice (no
+additional labels) is used. There's some padding involved here to make sure
+we're not flapping back and forth between these states.
+
+With this approach, EndpointSlices can be delivered everywhere (no additional
+labels), or to a zone (for-zone), or to a region (for-region). None of the
+proposed approaches involve a single Service having separate sets of
+EndpointSlices for each use case. As defined by [KEP
+1659](https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/1659-standard-topology-labels),
+"region" and "zone" are strictly hierarchical ("zones" are subsets of "regions")
+and zone names are unique across regions.
+
+### Backwards Compatibility
+We don't need create EndpointSlices without labels for backwards compatibility,
+we just need to ensure that consumers always support consuming EndpointSlices
+without these labels. Even if we updated the EndpointSlice controller to
+consistently label these EndpointSlices with `for-zone` or `for-region`, we
+couldn't guarantee that other producers would.
+
+There's nothing in any current consumer implementation that would break if
+additional labels like `for-zone` or `for-region` were added to EndpointSlices.
+All consumers will need to care about for this or the original approach is the
+`kubernetes.io/service-name` label. If they want to support subsetting, they can
+update their selectors as described in this KEP, but subsetting won't actually
+break any existing functionality.
```

Comment:
I think I don't fully agree with it.
It's not directly related to this KEP, because it only introduces the mechanism for subsetting. But once you get to subsetting, this is actually a broader problem than just endpoint-slice controller.

Imagine that you have HPA set up on a deployment (that is spanning 3 zones) and a frontend talking to that deployment running just in a single zone. So with subsetting, we will only be talking to the instances in the same zone, but will be scaling up in all zones (kind-of loosing capacity).

So I think that subsetting by itself may actually have very unexpected consequences to users (which we should mention here explicitly).
@thockin - FYI

I think I'm fine with this KEP on its own, but I would really like to be one of the aprovers for the KEP that will actually introduce some subsetting.

Reply:
Yep, you're completely right, we've already called out the scenario where all traffic originates from a single zone as being a potential failure point for that approach. I'll expand on that to include the HPA implications and add you as an approver for https://github.com/kubernetes/enhancements/pull/2005.

Reply:
ACK.

The only real solution to that problem is actual feedback, which we probably should consider.  I'm honestly hoping that subsetting gets us some way down that road for some large fraction of users, but I believe feedback will eventually be needed.

Reply:
I agree. What I was just saying is that:
(a) users should be aware of consequences if we decide to change defaults [not sure how to achieve it really] (i.e. subsetting by default)
(b) or ideally actually address all issues like the HPA-related one before making this a default (where feedback loop seem to be the only reasonable solution)